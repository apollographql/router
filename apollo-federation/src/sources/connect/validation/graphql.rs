//! Helper structs & functions for dealing with GraphQL schemas
use std::ops::Deref;

use apollo_compiler::Name;
use apollo_compiler::Schema;
use line_col::LineColLookup;

mod strings;

pub(super) use strings::GraphQLString;

pub(super) struct SchemaInfo<'schema> {
    pub(crate) schema: &'schema Schema,
    len: usize,
    lookup: LineColLookup<'schema>,
    pub(crate) connect_directive_name: &'schema Name,
    pub(crate) source_directive_name: &'schema Name,
}

impl<'schema> SchemaInfo<'schema> {
    pub(crate) fn new(
        schema: &'schema Schema,
        src: &'schema str,
        connect_directive_name: &'schema Name,
        source_directive_name: &'schema Name,
    ) -> Self {
        Self {
            schema,
            len: src.len(),
            lookup: LineColLookup::new(src),
            connect_directive_name,
            source_directive_name,
        }
    }

    /// Get the 1-based line and column values for an offset into this schema.
    ///
    /// # Returns
    /// The line and column, or `None` if the offset is not within the schema.
    pub(crate) fn line_col(&self, offset: usize) -> Option<(usize, usize)> {
        if offset > self.len {
            None
        } else {
            Some(self.lookup.get(offset))
        }
    }
}

impl Deref for SchemaInfo<'_> {
    type Target = Schema;

    fn deref(&self) -> &Self::Target {
        self.schema
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn line_col_lookup() {
        let src = r#"
            type Query {
                foo: String
            }
        "#;
        let schema = Schema::parse(src, "testSchema").unwrap();

        let name = "unused".try_into().unwrap();
        let schema_info = SchemaInfo::new(&schema, src, &name, &name);

        assert_eq!(schema_info.line_col(0), Some((1, 1)));
        assert_eq!(schema_info.line_col(4), Some((2, 4)));
        assert_eq!(schema_info.line_col(200), None);
    }
}
