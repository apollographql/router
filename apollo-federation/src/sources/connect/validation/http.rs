pub(super) mod headers;
pub(super) mod method;
pub(super) mod url;
