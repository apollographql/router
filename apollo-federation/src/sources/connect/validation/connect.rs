//! Parsing and validation of `@connect` directives

use apollo_compiler::Name;
use apollo_compiler::Node;
use apollo_compiler::ast::FieldDefinition;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::ObjectType;
use itertools::Itertools;

use self::entity::validate_entity_arg;
use self::selection::Selection;
use super::Code;
use super::Message;
use super::coordinates::ConnectDirectiveCoordinate;
use super::coordinates::connect_directive_name_coordinate;
use super::coordinates::source_name_value_coordinate;
use super::source::SourceName;
use crate::sources::connect::ConnectSpec;
use crate::sources::connect::id::ConnectedElement;
use crate::sources::connect::id::ObjectCategory;
use crate::sources::connect::spec::schema::CONNECT_SOURCE_ARGUMENT_NAME;
use crate::sources::connect::validation::connect::http::Http;
use crate::sources::connect::validation::graphql::SchemaInfo;

mod entity;
mod http;
mod selection;

pub(super) fn fields_seen_by_all_connects(
    schema: &SchemaInfo,
    all_source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    schema
        .types
        .values()
        .filter(|ty| !ty.is_built_in())
        .for_each(|extended_type| {
            if let ExtendedType::Object(node) = extended_type {
                match fields_seen_by_connectors_on_types(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }

                match fields_seen_by_connectors_on_fields(node, schema, all_source_names) {
                    Ok(fields) => seen_fields.extend(fields),
                    Err(errs) => messages.extend(errs),
                }
            }
        });

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on types are valid
fn fields_seen_by_connectors_on_types(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let connect_directives = object
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // TODO: find a better place for feature gates like this
    if schema.connect_link.spec == ConnectSpec::V0_1 {
        return Err(vec![Message {
            code: Code::FeatureUnavailable,
            message: format!(
                "Using `@{connect_directive_name}` on `type {object_name}` requires connectors v0.2. Learn more at https://go.apollo.dev/connectors/changelog.",
                object_name = object.name,
                connect_directive_name = schema.connect_directive_name(),
            ),
            locations: object
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        }]);
    }

    let mut messages = Vec::new();
    let mut seen_fields = Vec::new();

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Type { type_def: object },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                messages.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => messages.push(error),
        }

        // TODO: validate batch/this arguments as key fields

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    messages.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                messages.extend(errs);
                continue;
            }
        };
        messages.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

/// Make sure that any `@connect` directives on object fields are valid
fn fields_seen_by_connectors_on_fields(
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
    source_names: &[SourceName],
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let object_category = if schema
        .schema_definition
        .query
        .as_ref()
        .is_some_and(|query| query.name == object.name)
    {
        ObjectCategory::Query
    } else if schema
        .schema_definition
        .mutation
        .as_ref()
        .is_some_and(|mutation| mutation.name == object.name)
    {
        ObjectCategory::Mutation
    } else {
        ObjectCategory::Other
    };
    let mut seen_fields = Vec::new();
    let mut messages = Vec::new();
    for field in object.fields.values() {
        match fields_seen_by_connector(field, object_category, source_names, object, schema) {
            Ok(fields) => seen_fields.extend(fields),
            Err(errs) => messages.extend(errs),
        }
    }
    if messages.is_empty() {
        Ok(seen_fields)
    } else {
        Err(messages)
    }
}

fn fields_seen_by_connector(
    field: &Component<FieldDefinition>,
    category: ObjectCategory,
    source_names: &[SourceName],
    object: &Node<ObjectType>,
    schema: &SchemaInfo,
) -> Result<Vec<(Name, Name)>, Vec<Message>> {
    let source_map = &schema.sources;
    let mut errors = Vec::new();
    let connect_directives = field
        .directives
        .iter()
        .filter(|directive| directive.name == *schema.connect_directive_name())
        .collect_vec();

    if connect_directives.is_empty() {
        return Ok(Vec::new());
    }

    // mark the field with a @connect directive as seen
    let mut seen_fields = vec![(object.name.clone(), field.name.clone())];

    // direct recursion isn't allowed, like a connector on User.friends: [User]
    if &object.name == field.ty.inner_named_type() {
        errors.push(Message {
            code: Code::CircularReference,
            message: format!(
                "Direct circular reference detected in `{}.{}: {}`. For more information, see https://go.apollo.dev/connectors/limitations#circular-references",
                object.name,
                field.name,
                field.ty
            ),
            locations: field.line_column_range(source_map).into_iter().collect(),
        });
    }

    for connect_directive in connect_directives {
        let coordinate = ConnectDirectiveCoordinate {
            directive: connect_directive,
            element: ConnectedElement::Field {
                parent_type: object,
                parent_category: category,
                field_def: field,
            },
        };

        let selection = match Selection::parse(coordinate, schema) {
            Ok(selection) => selection,
            Err(err) => {
                errors.push(err);
                continue;
            }
        };
        match selection.type_check(schema) {
            Ok(seen) => seen_fields.extend(seen),
            Err(error) => errors.push(error),
        }

        errors
            .extend(validate_entity_arg(field, connect_directive, object, schema, category).err());

        let source_name =
            match validate_source_name(connect_directive, &coordinate, source_names, schema) {
                Ok(source_name) => source_name,
                Err(err) => {
                    errors.push(err);
                    continue;
                }
            };

        // TODO: Do all parsing in one stage, then all type checking in a later stage
        let http = match Http::parse(coordinate, source_name, schema) {
            Ok(http) => http,
            Err(errs) => {
                errors.extend(errs);
                continue;
            }
        };
        errors.extend(http.type_check(schema).err().into_iter().flatten());
    }

    if errors.is_empty() {
        Ok(seen_fields)
    } else {
        Err(errors)
    }
}

pub(super) fn validate_source_name<'schema>(
    directive: &Node<Directive>,
    coordinate: &ConnectDirectiveCoordinate,
    source_names: &'schema [SourceName],
    schema: &SchemaInfo,
) -> Result<Option<&'schema SourceName<'schema>>, Message> {
    let Some(source_name_arg) = directive
        .arguments
        .iter()
        .find(|arg| arg.name == CONNECT_SOURCE_ARGUMENT_NAME)
    else {
        return Ok(None);
    };

    let resolved_source_name = source_names
        .iter()
        .find(|name| **name == source_name_arg.value);

    if let Some(source_name) = resolved_source_name {
        return Ok(Some(source_name));
    }
    // A source name was set but doesn't match a defined source
    // TODO: Pick a suggestion that's not just the first defined source
    let qualified_directive = connect_directive_name_coordinate(
        schema.connect_directive_name(),
        &source_name_arg.value,
        coordinate,
    );
    if let Some(first_source_name) = source_names.first() {
        Err(Message {
            code: Code::SourceNameMismatch,
            message: format!(
                "{qualified_directive} does not match any defined sources. Did you mean \"{first_source_name}\"?",
                first_source_name = first_source_name.as_str(),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    } else {
        Err(Message {
            code: Code::NoSourcesDefined,
            message: format!(
                "{qualified_directive} specifies a source, but none are defined. Try adding {coordinate} to the schema.",
                coordinate = source_name_value_coordinate(
                    schema.source_directive_name(),
                    &source_name_arg.value
                ),
            ),
            locations: source_name_arg
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    }
}
