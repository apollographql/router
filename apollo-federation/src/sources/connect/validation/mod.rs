//! Validation of the `@source` and `@connect` directives.

// No panics allowed in this module
#![cfg_attr(
    not(test),
    deny(
        clippy::exit,
        clippy::panic,
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::indexing_slicing,
        clippy::unimplemented,
        clippy::todo
    )
)]

mod coordinates;
mod entity;
mod extended_type;
mod graphql;
mod http;
mod selection;
mod source_name;
mod variable;

use std::collections::HashMap;
use std::fmt::Display;
use std::ops::Range;

use apollo_compiler::ast::OperationType;
use apollo_compiler::ast::Value;
use apollo_compiler::collections::IndexSet;
use apollo_compiler::executable::FieldSet;
use apollo_compiler::name;
use apollo_compiler::parser::LineColumn;
use apollo_compiler::parser::Parser;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::Directive;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::ObjectType;
use apollo_compiler::validation::Valid;
use apollo_compiler::Name;
use apollo_compiler::Node;
use apollo_compiler::Schema;
use coordinates::source_http_argument_coordinate;
use entity::field_set_error;
use entity::EntityKeyChecker;
use extended_type::validate_extended_type;
use itertools::Itertools;
use source_name::SourceName;
use strum_macros::Display;
use strum_macros::IntoStaticStr;
use url::Url;

use super::Connector;
use crate::link::federation_spec_definition::FEDERATION_FIELDS_ARGUMENT_NAME;
use crate::link::federation_spec_definition::FEDERATION_KEY_DIRECTIVE_NAME_IN_SPEC;
use crate::link::federation_spec_definition::FEDERATION_RESOLVABLE_ARGUMENT_NAME;
use crate::link::spec::Identity;
use crate::link::Import;
use crate::link::Link;
use crate::sources::connect::spec::schema::HTTP_ARGUMENT_NAME;
use crate::sources::connect::spec::schema::SOURCE_BASE_URL_ARGUMENT_NAME;
use crate::sources::connect::spec::schema::SOURCE_DIRECTIVE_NAME_IN_SPEC;
use crate::sources::connect::spec::schema::SOURCE_NAME_ARGUMENT_NAME;
use crate::sources::connect::validation::coordinates::BaseUrlCoordinate;
use crate::sources::connect::validation::coordinates::HttpHeadersCoordinate;
use crate::sources::connect::validation::graphql::GraphQLString;
use crate::sources::connect::validation::graphql::SchemaInfo;
use crate::sources::connect::validation::http::headers;
use crate::sources::connect::ConnectSpec;
use crate::subgraph::spec::CONTEXT_DIRECTIVE_NAME;
use crate::subgraph::spec::EXTERNAL_DIRECTIVE_NAME;
use crate::subgraph::spec::FROM_CONTEXT_DIRECTIVE_NAME;

/// Validate the connectors-related directives `@source` and `@connect`.
///
/// This function attempts to collect as many validation errors as possible, so it does not bail
/// out as soon as it encounters one.
pub fn validate(source_text: &str, file_name: &str) -> Vec<Message> {
    // TODO: Use parse_and_validate (adding in directives as needed)
    // TODO: Handle schema errors rather than relying on JavaScript to catch it later
    let schema = Schema::parse(source_text, file_name)
        .unwrap_or_else(|schema_with_errors| schema_with_errors.partial);
    let connect_identity = ConnectSpec::identity();
    let Some((link, link_directive)) = Link::for_identity(&schema, &connect_identity) else {
        return Vec::new(); // There are no connectors-related directives to validate
    };

    let federation = Link::for_identity(&schema, &Identity::federation_identity());
    let external_directive_name = federation
        .map(|(link, _)| link.directive_name_in_schema(&EXTERNAL_DIRECTIVE_NAME))
        .unwrap_or(EXTERNAL_DIRECTIVE_NAME.clone());

    let mut messages = check_conflicting_directives(&schema);

    let source_directive_name = ConnectSpec::source_directive_name(&link);
    let connect_directive_name = ConnectSpec::connect_directive_name(&link);
    let schema_info = SchemaInfo::new(
        &schema,
        source_text,
        &connect_directive_name,
        &source_directive_name,
    );

    let source_directives: Vec<SourceDirective> = schema
        .schema_definition
        .directives
        .iter()
        .filter(|directive| directive.name == source_directive_name)
        .map(|directive| validate_source(directive, &schema_info))
        .collect();

    let mut valid_source_names = HashMap::new();
    let all_source_names = source_directives
        .iter()
        .map(|directive| directive.name.clone())
        .collect_vec();
    for directive in source_directives {
        messages.extend(directive.errors);
        match directive.name.into_value_or_error(&schema_info.sources) {
            Err(error) => messages.push(error),
            Ok(name) => valid_source_names
                .entry(name)
                .or_insert_with(Vec::new)
                .extend(
                    directive
                        .directive
                        .node
                        .line_column_range(&schema_info.sources),
                ),
        }
    }
    for (name, locations) in valid_source_names {
        if locations.len() > 1 {
            messages.push(Message {
                message: format!("Every `@{source_directive_name}({SOURCE_NAME_ARGUMENT_NAME}:)` must be unique. Found duplicate name {name}."),
                code: Code::DuplicateSourceName,
                locations,
            });
        }
    }

    let mut seen_fields = IndexSet::default();

    let connect_errors = schema.types.values().flat_map(|extended_type| {
        validate_extended_type(
            extended_type,
            &schema_info,
            &all_source_names,
            &mut seen_fields,
        )
    });
    messages.extend(connect_errors);

    if source_directive_name == DEFAULT_SOURCE_DIRECTIVE_NAME
        && messages
            .iter()
            .any(|error| error.code == Code::NoSourcesDefined)
    {
        messages.push(Message {
            code: Code::NoSourceImport,
            message: format!("The `@{SOURCE_DIRECTIVE_NAME_IN_SPEC}` directive is not imported. Try adding `@{SOURCE_DIRECTIVE_NAME_IN_SPEC}` to `import` for `@{link_name}(url: \"{connect_identity}\")`", link_name=link_directive.name),
            locations: link_directive.line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        });
    }

    if should_do_advanced_validations(&messages) {
        messages.extend(check_seen_fields(
            &schema_info,
            &seen_fields,
            &external_directive_name,
        ));

        match advanced_validations(&schema, file_name, ConnectSpec::V0_1) {
            Ok(multiple) => messages.extend(multiple),
            Err(Some(single)) => messages.push(single),
            _ => {} // let the rest of composition handle this
        };
    }

    messages
}

fn advanced_validations(
    schema: &Schema,
    subgraph_name: &str,
    spec: ConnectSpec,
) -> Result<Vec<Message>, Option<Message>> {
    let mut messages = vec![];

    let connectors = Connector::from_schema(schema, subgraph_name, spec).map_err(|_| None)?;

    let mut entity_checker = EntityKeyChecker::default();

    for (field_set, directive) in find_all_resolvable_keys(schema) {
        entity_checker.add_key(&field_set, directive);
    }

    for (_, connector) in connectors {
        if let Some(field_set) = connector.resolvable_key(schema).map_err(|_| {
            let variables = connector.variable_references().collect_vec();
            field_set_error(&variables, connector.id.directive.field.type_name())
        })? {
            entity_checker.add_connector(field_set);
        }
    }

    messages.extend(entity_checker.check_for_missing_entity_connectors(schema));

    Ok(messages)
}

/// We'll avoid doing this work if there are bigger issues with the schema.
/// Otherwise we might emit a large number of diagnostics that will
/// distract from the main problems.
fn should_do_advanced_validations(messages: &[Message]) -> bool {
    messages.is_empty()
}

/// Check that all fields defined in the schema are resolved by a connector.
fn check_seen_fields(
    schema: &SchemaInfo,
    seen_fields: &IndexSet<(Name, Name)>,
    external_directive_name: &Name,
) -> Vec<Message> {
    let all_fields: IndexSet<_> = schema
        .types
        .values()
        .filter_map(|extended_type| {
            if extended_type.is_built_in() {
                return None;
            }
            // ignore root fields, we have different validations for them
            if schema.root_operation(OperationType::Query) == Some(extended_type.name())
                || schema.root_operation(OperationType::Mutation) == Some(extended_type.name())
                || schema.root_operation(OperationType::Subscription) == Some(extended_type.name())
            {
                return None;
            }
            let coord = |(name, _): (&Name, _)| (extended_type.name().clone(), name.clone());
            match extended_type {
                ExtendedType::Object(object) => Some(
                    // ignore @external fields
                    object
                        .fields
                        .iter()
                        .filter(|(_, def)| {
                            !def.directives
                                .iter()
                                .any(|dir| &dir.name == external_directive_name)
                        })
                        .map(coord),
                ),
                ExtendedType::Interface(_) => None, // TODO: when interfaces are supported (probably should include fields from implementing/member types as well)
                _ => None,
            }
        })
        .flatten()
        .collect();

    (&all_fields - seen_fields).iter().map(move |(parent_type, field_name)| {
        let Ok(field_def) = schema.type_field(parent_type, field_name) else {
            // This should never happen, but if it does, we don't want to panic
            return Message {
                code: Code::GraphQLError,
                message: format!(
                    "Field `{parent_type}.{field_name}` is missing from the schema.",
                ),
                locations: Vec::new(),
            };
        };
        Message {
            code: Code::ConnectorsUnresolvedField,
            message: format!(
                "No connector resolves field `{parent_type}.{field_name}`. It must have a `@{connect_directive_name}` directive or appear in `@{connect_directive_name}(selection:)`.",
                connect_directive_name = schema.connect_directive_name
            ),
            locations: field_def.line_column_range(&schema.sources).into_iter().collect(),
        }
    }).collect()
}

fn check_conflicting_directives(schema: &Schema) -> Vec<Message> {
    let Some((fed_link, fed_link_directive)) =
        Link::for_identity(schema, &Identity::federation_identity())
    else {
        return Vec::new();
    };

    // TODO: make the `Link` code retain locations directly instead of reparsing stuff for validation
    let imports = fed_link_directive
        .specified_argument_by_name(&name!("import"))
        .and_then(|arg| arg.as_list())
        .into_iter()
        .flatten()
        .filter_map(|value| Import::from_value(value).ok().map(|import| (value, import)))
        .collect_vec();

    let disallowed_imports = [CONTEXT_DIRECTIVE_NAME, FROM_CONTEXT_DIRECTIVE_NAME];
    fed_link
        .imports
        .into_iter()
        .filter_map(|import| {
            disallowed_imports
                .contains(&import.element)
                .then(|| Message {
                    code: Code::ConnectorsUnsupportedFederationDirective,
                    message: format!(
                        "The directive `@{import}` is not supported when using connectors.",
                        import = import.alias.as_ref().unwrap_or(&import.element)
                    ),
                    locations: imports
                        .iter()
                        .find_map(|(value, reparsed)| {
                            (*reparsed == *import).then(|| value.line_column_range(&schema.sources))
                        })
                        .flatten()
                        .into_iter()
                        .collect(),
                })
        })
        .collect()
}

const DEFAULT_SOURCE_DIRECTIVE_NAME: &str = "connect__source";
#[allow(unused)]
const DEFAULT_CONNECT_DIRECTIVE_NAME: &str = "connect__connect";

fn validate_source(directive: &Component<Directive>, schema: &SchemaInfo) -> SourceDirective {
    let name = SourceName::from_directive(directive);
    let mut errors = Vec::new();

    if let Some(http_arg) = directive
        .specified_argument_by_name(&HTTP_ARGUMENT_NAME)
        .and_then(|arg| arg.as_object())
    {
        // Validate URL argument
        if let Some(url_value) = http_arg
            .iter()
            .find_map(|(key, value)| (key == &SOURCE_BASE_URL_ARGUMENT_NAME).then_some(value))
        {
            if let Some(url_error) = parse_url(
                url_value,
                BaseUrlCoordinate {
                    source_directive_name: &directive.name,
                },
                schema,
            )
            .err()
            {
                errors.push(url_error);
            }
        }

        errors.extend(headers::validate_arg(
            http_arg,
            schema,
            HttpHeadersCoordinate::Source {
                directive_name: &directive.name,
            },
        ));
    } else {
        errors.push(Message {
            code: Code::GraphQLError,
            message: format!(
                "{coordinate} must have a `{HTTP_ARGUMENT_NAME}` argument.",
                coordinate = source_http_argument_coordinate(&directive.name),
            ),
            locations: directive
                .line_column_range(&schema.sources)
                .into_iter()
                .collect(),
        })
    }

    SourceDirective {
        name,
        errors,
        directive: directive.clone(),
    }
}

/// A `@source` directive along with any errors related to it.
struct SourceDirective {
    name: SourceName,
    errors: Vec<Message>,
    directive: Component<Directive>,
}

fn parse_url<Coordinate: Display + Copy>(
    value: &Node<Value>,
    coordinate: Coordinate,
    schema: &SchemaInfo,
) -> Result<(), Message> {
    let str_value = GraphQLString::new(value, &schema.sources).map_err(|_| Message {
        code: Code::GraphQLError,
        message: format!("The value for {coordinate} must be a string."),
        locations: value
            .line_column_range(&schema.sources)
            .into_iter()
            .collect(),
    })?;
    let url = Url::parse(str_value.as_str()).map_err(|inner| Message {
        code: Code::InvalidUrl,
        message: format!("The value {value} for {coordinate} is not a valid URL: {inner}."),
        locations: value
            .line_column_range(&schema.sources)
            .into_iter()
            .collect(),
    })?;
    http::url::validate_base_url(&url, coordinate, value, str_value, schema)
}

/// For an object type, get all the keys (and directive nodes) that are resolvable.
///
/// The FieldSet returned here is what goes in the `fields` argument, so `id` in `@key(fields: "id")`
fn resolvable_key_fields<'a>(
    object: &'a Node<ObjectType>,
    schema: &'a Schema,
) -> impl Iterator<Item = (FieldSet, &'a Component<Directive>)> + 'a {
    object
        .directives
        .iter()
        .filter(|directive| directive.name == FEDERATION_KEY_DIRECTIVE_NAME_IN_SPEC)
        .filter(|directive| {
            directive
                .arguments
                .iter()
                .find(|arg| arg.name == FEDERATION_RESOLVABLE_ARGUMENT_NAME)
                .and_then(|arg| arg.value.to_bool())
                .unwrap_or(true)
        })
        .filter_map(|directive| {
            if let Some(fields_str) = directive
                .arguments
                .iter()
                .find(|arg| arg.name == FEDERATION_FIELDS_ARGUMENT_NAME)
                .map(|arg| &arg.value)
                .and_then(|value| value.as_str())
            {
                Parser::new()
                    .parse_field_set(
                        Valid::assume_valid_ref(schema),
                        object.name.clone(),
                        fields_str.to_string(),
                        "",
                    )
                    .ok()
                    .map(|field_set| (field_set, directive))
            } else {
                None
            }
        })
}

fn find_all_resolvable_keys(schema: &Schema) -> Vec<(FieldSet, &Component<Directive>)> {
    schema
        .types
        .values()
        .flat_map(|extended_type| match extended_type {
            ExtendedType::Object(object) => Some(resolvable_key_fields(object, schema)),
            _ => None,
        })
        .flatten()
        .collect()
}

type DirectiveName = Name;

#[derive(Debug)]
pub struct Message {
    /// A unique, per-error code to allow consuming tools to take specific actions. These codes
    /// should not change once stabilized.
    pub code: Code,
    /// A human-readable message describing the error. These messages are not stable, tools should
    /// not rely on them remaining the same.
    ///
    /// # Formatting messages
    /// 1. Messages should be complete sentences, starting with capitalization as appropriate and
    ///    ending with punctuation.
    /// 2. When referring to elements of the schema, use
    ///    [schema coordinates](https://github.com/graphql/graphql-wg/blob/main/rfcs/SchemaCoordinates.md)
    ///    with any additional information added as required for clarity (e.g., the value of an arg).
    /// 3. When referring to code elements (including schema coordinates), surround them with
    ///    backticks. This clarifies that `Type.field` is not ending a sentence with its period.
    pub message: String,
    pub locations: Vec<Range<LineColumn>>,
}

/// The error code that will be shown to users when a validation fails during composition.
///
/// Note that these codes are global, not scoped to connectors, so they should attempt to be
/// unique across all pieces of composition, including JavaScript components.
#[derive(Clone, Copy, Debug, Display, Eq, IntoStaticStr, PartialEq)]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
pub enum Code {
    /// A problem with GraphQL syntax or semantics was found. These will usually be caught before
    /// this validation process.
    GraphQLError,
    DuplicateSourceName,
    InvalidSourceName,
    EmptySourceName,
    /// A provided URL was not valid
    InvalidUrl,
    /// A URL scheme is not `http` or `https`
    InvalidUrlScheme,
    SourceNameMismatch,
    SubscriptionInConnectors,
    /// Query field is missing the `@connect` directive
    QueryFieldMissingConnect,
    /// Mutation field is missing the `@connect` directive
    MutationFieldMissingConnect,
    /// The `@connect` is using a `source`, but the URL is absolute. This is trouble because
    /// the `@source` URL will be joined with the `@connect` URL, so the `@connect` URL should
    /// actually be a path only.
    AbsoluteConnectUrlWithSource,
    /// The `@connect` directive is using a relative URL (path only) but does not define a `source`.
    /// This is just a specialization of [`Self::InvalidUrl`] that provides a better suggestion for
    /// the user.
    RelativeConnectUrlWithoutSource,
    /// This is a specialization of [`Self::SourceNameMismatch`] that provides a better suggestion.
    NoSourcesDefined,
    /// The subgraph doesn't import the `@source` directive. This isn't necessarily a problem, but
    /// is likely a mistake.
    NoSourceImport,
    /// The `@connect` directive has multiple HTTP methods when only one is allowed.
    MultipleHttpMethods,
    /// The `@connect` directive is missing an HTTP method.
    MissingHttpMethod,
    /// The `entity` argument should only be used on the root `Query` field.
    EntityNotOnRootQuery,
    /// The arguments to the entity reference resolver do not match the entity type.
    EntityResolverArgumentMismatch,
    /// The `entity` argument should only be used with non-list, nullable, object types.
    EntityTypeInvalid,
    /// A @key is defined without a cooresponding entity connector.
    MissingEntityConnector,
    /// A syntax error in `selection`
    InvalidJsonSelection,
    /// A cycle was detected within a `selection`
    CircularReference,
    /// A field was selected but is not defined on the type
    SelectedFieldNotFound,
    /// A group selection (`a { b }`) was used, but the field is not an object
    GroupSelectionIsNotObject,
    /// The `name` mapping must be unique for all headers.
    HttpHeaderNameCollision,
    /// A provided header in `@source` or `@connect` was not valid
    InvalidHeader,
    /// Certain directives are not allowed when using connectors
    ConnectorsUnsupportedFederationDirective,
    /// Abstract types are not allowed when using connectors
    ConnectorsUnsupportedAbstractType,
    /// Fields that return an object type must use a group JSONSelection `{}`
    GroupSelectionRequiredForObject,
    /// Fields in the schema that aren't resolved by a connector
    ConnectorsUnresolvedField,
    /// A field resolved by a connector has arguments defined
    ConnectorsFieldWithArguments,
    /// Part of the `@connect` refers to an `$args` which is not defined
    UndefinedArgument,
    /// Part of the `@connect` refers to an `$this` which is not defined
    UndefinedField,
    /// A type used in a variable is not yet supported (i.e., unions)
    UnsupportedVariableType,
    /// A variable is nullable in a location which requires non-null at runtime
    NullabilityMismatch,
}

impl Code {
    pub const fn severity(&self) -> Severity {
        match self {
            Self::NoSourceImport | Self::NullabilityMismatch => Severity::Warning,
            _ => Severity::Error,
        }
    }
}

/// Given the [`Code`] of a [`Message`], how important is that message?
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Severity {
    /// This is an error, validation as failed.
    Error,
    /// The user probably wants to know about this, but it doesn't halt composition.
    Warning,
}

#[cfg(test)]
mod test_validate_source {
    use std::fs::read_to_string;

    use insta::assert_snapshot;
    use insta::glob;

    use super::*;

    #[test]
    fn validation_tests() {
        insta::with_settings!({prepend_module_to_snapshot => false}, {
            glob!("test_data", "**/*.graphql", |path| {
                let schema = read_to_string(path).unwrap();
                let errors = validate(&schema, path.to_str().unwrap());
                assert_snapshot!(format!("{:#?}", errors));
            });
        });
    }
}
