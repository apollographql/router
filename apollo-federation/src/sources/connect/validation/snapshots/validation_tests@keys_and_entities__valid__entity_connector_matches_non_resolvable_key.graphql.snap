---
source: apollo-federation/src/sources/connect/validation/mod.rs
expression: "format!(\"{:#?}\", errors)"
input_file: apollo-federation/src/sources/connect/validation/test_data/keys_and_entities/valid/entity_connector_matches_non_resolvable_key.graphql
---
[]
